### PR TITLE
nixos: better tor config

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change
SLNOS had a (IMHO) superior version of #28081 for half a year already. I rebased it onto NixOS master. Feel free to bikeshed and reject :)

###### Things done
- [X] It works
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bjornfor** [comment]
> SLNOS had a (IMHO) superior version of #28081 for half a year already.

Where is the SLNOS code/fork? I don't see it on github.com nor from internet search.

**Comment by oxij** [comment]
> Where is the SLNOS code/fork? I don't see it on github.com nor from internet search. 

That's a good question. Currently we abuse BSD license by not sharing
the code anywhere except between members of our group.

It will probably go public on https://github.com/SLNOS after s6 init
becomes usable.

**Comment by oxij** [comment]
Travis error is a doc build error which is a ghc build error.

This patchset had ~6 months of testing, so lets proceed to the discussion.

**Comment by tnias** [comment]
I like the descriptive texts a lot.

File: nixos/modules/services/security/tor.nix [hunk 1]
```diff
@@ -330,79 +458,123 @@ in
...
+
+                   toHost = mkOption {
+                     type = types.str;
+                     default = "127.0.0.1";
+                     description = "Mapping destination host.";
+                   };
+
+                   toPort = mkOption {
+                     type = types.str;
+                     defaultText = "port";
```
Comment (fpletz) [inline nixos/modules/services/security/tor.nix]: This should be unnecessary, right?

**Review by fpletz — APPROVED** [review]
I like this version and especially the docs but the duplicated effort on the hidden services implementation is a pity. I would welcome you open sourcing your fork even though the license doesn't dictate that.

**Comment by oxij** [comment]
Franz Pletz <notifications@github.com> writes:

> This should be unnecessary, right?

Yep, probably a leftover of another version or something.

Also fixed another tiny type bug I haven't noticed before.

> I like this version and especially the docs but the duplicated effort on the hidden services implementation is a pity. I would welcome you open sourcing your fork even though the license doesn't dictate that.

Yeah, but the problem is that its ~200 patches over master ATM and most
of them depend on (or are pretty useless without) changes that were
rejected from master or our custom tooling (we don't support
`nixos-rebuild`). Splitting those patches into "this can go upstream"
and "this is only for our space-age technology" takes a lot of effort
nobody is willing to put in unless things start conflicting with master.

Also, many SLNOS patchsets are not beautiful, which is an (expected)
consequence of out "Is this an actual problem? Yes? Do you have a better
solution for it right now? No? Stop bikeshedding and merge this" policy.

Also note that we can't just publish all of the patches and have the
community fix them as public SLNOS members (currently only me) have to
fix code style and rewrite all the doc strings and comments as per our
stylometry policy. This takes lots of effort (and sometimes some
research as I can't document things I don't understand).

If something gets an equivalent or better implementation in master while
our patches ferment we just drop our patches or reimplement features we
want on top of whatever new master got. This happens quite regularly but
since SLNOS cares about features, not credit, we are ok with that. In
this instance I just loved the SLNOS version of `options` so much (and
finished doc strings awhile ago; but those patches got lost in the sea
of other stuff and I remembered about them only on merge conflict with
master) I felt it would be a loss to just drop them.

In short, I don't know a good way to solve the duplicated effort part.
If you have any suggestions I would welcome them.

At least in one instance other developer read my nix-dev announcement
and wrote an e-mail of "I want to implement X which is on SLNOS TODO
list. Does SLNOS have that already by any chance or should I proceed
implementing it myself?" nature. Which was very nice of him/her. I, too,
put _all_ NixOS github notifications and nix-dev into `notmuch` and
search that and `git log` (sometimes the future is in the reverted past)
before doing _or discussing_ any substantial changes to nixpkgs or
SLNOS. But how many other people do that on a regular basis? You can't
do that without using space-age technology (xapian) to handle your
e-mail and most people outsource their e-mail to people that use
space-age technology to make your e-mail easy to mine, not to make huge
streams of new e-mail and enormous e-mail archives you already have easy
to consume.

(If you're reading this and don't use `notmuch` go check it out,
seriously. This planet has two pieces of software that I don't regret
putting a ton of effort to learn to use and automate: Emacs (org-mode)
and notmuch. It also works perfectly with Emacs and, after some
`org-capture` magic, with org-mode. But if you think `notmuch` is a
little (not) too much for you then check `mailpile` out, it has most of
space-age technology of `notmuch` within a conventional UX your mom
could work with.)

**Comment by oxij** [comment]
Also dropped some dead code.

**Comment by oxij** [comment]
Tested the fixes myself. Travis is happy too.

**Comment by oxij** [comment]
Yay! Thanks.

**Comment by rycee** [comment]
I think this should have been in the release notes as a backwards incompatibility. When updating my Tor relay today from 17.03 I got
```
error: The option `services.tor.relay.role' is used but not defined.
```

Wasn't difficult to fix by adding a `role = "relay"` to my configuration but it took me by surprise.

**Comment by oxij** [comment]
> I think this should have been in the release notes as a backwards incompatibility. When updating my Tor relay today from 17.03 I got
> ```
> error: The option `services.tor.relay.role' is used but not defined.
> ```
>
> Wasn't difficult to fix by adding a `role = "relay"` to my configuration but it took me by surprise.

It's not an incompatibility, it was put there intentionally to make you
think twice before enabling anything relay when you enable it the first
time and to make you review your `services.tor` config when you had it
enabled before (since a lot has changed). See the doc strings and Dmitry
Bogatov's case.

**Pull request merged**